### ISSUE TITLE
Restore doesn't restore dependent children

### ISSUE THREAD

**Issue by cj** [rubysherpas/paranoia#20]
Using `.restore!` doesn't restore the dependent children.

```
1.9.3-p0 :008 > owner = Owner.unscoped.first
  DEBUG - Owner Load (0.7ms)  SELECT `owners`.* FROM `owners` LIMIT 1
 => #<Owner id: 2, first_name: "CJ", last_name: "Lazell", phone: nil, creator_id: nil, updater_id: nil, deleted_at: "2012-02-17 14:30:26", created_at: "2012-02-17 14:28:45", updated_at: "2012-02-17 14:30:26"> 
1.9.3-p0 :009 > owner.restore!
  DEBUG -  (10.4ms)  BEGIN
  DEBUG -  (0.8ms)  UPDATE `owners` SET `deleted_at` = NULL, `updated_at` = '2012-02-17 14:36:33' WHERE `owners`.`id` = 2
  DEBUG -  (3.0ms)  COMMIT
 => true 
```

Also if you call `.restore!` on a just destroyed record you get an error:

```
1.9.3-p0 :001 > owner = Owner.first
  DEBUG - Owner Load (0.2ms)  SELECT `owners`.* FROM `owners` WHERE `owners`.`deleted_at` IS NULL LIMIT 1
 => #<Owner id: 2, first_name: "CJ", last_name: "Lazell", phone: nil, creator_id: nil, updater_id: nil, deleted_at: nil, created_at: "2012-02-17 14:28:45", updated_at: "2012-02-17 14:28:45"> 
1.9.3-p0 :002 > owner.destroy
  DEBUG - Address Load (3.7ms)  SELECT `addresses`.* FROM `addresses` WHERE `addresses`.`deleted_at` IS NULL AND `addresses`.`location_id` = 2 AND `addresses`.`location_type` = 'Owner' LIMIT 1
  DEBUG -  (0.1ms)  BEGIN
  DEBUG -  (2.0ms)  UPDATE `addresses` SET `deleted_at` = '2012-02-17 14:30:26', `updated_at` = '2012-02-17 14:30:26' WHERE `addresses`.`id` = 1
  DEBUG -  (1.7ms)  COMMIT
  DEBUG -  (0.1ms)  BEGIN
  DEBUG -  (3.4ms)  UPDATE `owners` SET `deleted_at` = '2012-02-17 14:30:26', `updated_at` = '2012-02-17 14:30:26' WHERE `owners`.`id` = 2
  DEBUG -  (0.5ms)  COMMIT
 => #<Owner id: 2, first_name: "CJ", last_name: "Lazell", phone: nil, creator_id: nil, updater_id: nil, deleted_at: "2012-02-17 14:30:26", created_at: "2012-02-17 14:28:45", updated_at: "2012-02-17 14:30:26"> 
1.9.3-p0 :003 > owner.restore!
RuntimeError: can't modify frozen Hash
```

**Comment by bgadoury** [comment]
CJ,

I have a patch for automatically restoring dependent children where the parent has declared has_many or has_one associations, the associations use :dependent => (:destroy|:delete), and the children classes are also .paranoid? == true. I'm writing unit tests for my fork now, and will make a pull request when the test coverage is acceptable.

As for calling .restore! on an instance you just called destroy on and getting that error, that's due to Paranoia's destroy and delete freezing the model instance just like ActiveRecord does. Behaving in a manner consistent with ActiveRecord is probably more important than making it less aggravating to debug and test in IRB.

**Comment by arsduo** [comment]
Hi @bgadoury, did you ever finish your restorying-the-dependent-children patch?  If not I'd be happy to try finishing it.  It'd be really useful for us.

Thanks!

**Comment by findchris** [comment]
+1

**Comment by bgadoury** [comment]
@arsduo, I'm sorry but I just saw your message. I put it on the shelf before I got any tests written, but I remember it being pretty messy.

I'll try to find some time to dig it up and see how I left it.

**Comment by mwalsher** [comment]
+1, this would be really useful.

**Comment by radar** [comment]
If you would like this feature, please spend some time investigating a patch for it.

**Comment by tsai146** [comment]
Is #91 fix your issue?

**Comment by bgadoury** [comment]
I didn't create this issue, but #91 looks like a great solution.

**Comment by radar** [comment]
Fixed by #91.
